pkg/utils/kubernetes/kubernetes.go

```diff
@@ -513,13 +513,26 @@ func NewestPodForDeployment(ctx context.Context, c client.Reader, deployment *ap
 		return nil, fmt.Errorf("object is not of type *appsv1.ReplicaSet but %T", replicaSet)
 	}
 
+	if newestReplicaSet.Spec.Selector == nil {
+		return nil, fmt.Errorf("no pod selector specified in replicaSet %s/%s", newestReplicaSet.Namespace, newestReplicaSet.Name)
+	}
+
+	if len(newestReplicaSet.Spec.Selector.MatchLabels)+len(newestReplicaSet.Spec.Selector.MatchExpressions) == 0 {
+		return nil, fmt.Errorf("no matchLabels or matchExpressions specified in replicaSet %s/%s", newestReplicaSet.Namespace, newestReplicaSet.Name)
+	}
+
+	podSelector, err := metav1.LabelSelectorAsSelector(newestReplicaSet.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert the pod selector from ReplicaSet %s/%s: %v", newestReplicaSet.Namespace, newestReplicaSet.Name, err)
+	}
+
+	listOpts = append(listOpts, client.MatchingLabelsSelector{Selector: podSelector})
```

Comment:
Well, IIUC this optimization is done to select only the Pods that belong to the  newest ReplicaSet, right?
I think that we won't gain much from such optimization as we already filter out by the Deployment matchLabels, so in worst case it will list few Pods more than without this optimization.
To be honest I vote to keep the old way as it seems to me much more elegant, easy to read && understand. Do you agree?

Reply:
If accept the replica set always use the `newestReplicaSet.Spec.Selector.MatchLabels` and never `newestReplicaSet.Spec.Selector.MatchExpressions`, as it is already done for the Deployment, it will be also very easy to read. 
I think it can be simplified even now by:
```golang
listOpts = append(listOpts, client.MatchingLabels(map[string]string{"pod-template-hash": newestReplicaSet.Name}))
```
but I suppose this is not defined in the API and can be changed in any future k8s version. 

I do not like the ownerReference filtering because it is done on the client side when it can be done on the server side with the label selector. 

So I am open to simplify the label selector but really would like to remove the ownerRef filtering.

Reply:
> [...] but really would like to remove the ownerRef filtering.

Such filtering is already in place few lines above for ReplicaSets.

I don't want to repeat myself with https://github.com/gardener/gardener/pull/4050#discussion_r632368582. I would rather vote to have the old logic in place (that can list for example 6 Pods, instead of 3). This code path is already being executed in special cases.
Is it possible to have a separate PR for this optimization where we can get a`second-opinion`? In this way we can proceed with this PR to fix corresponding issue?

Reply:
> Is it possible to have a separate PR for this optimization where we can get asecond-opinion? In this way we can proceed with this PR to fix corresponding issue

Sure.